kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/ColumnMapping.java

```diff
@@ -170,6 +170,194 @@ static int findMaxColumnId(StructType schema) {
     return maxColumnId;
   }
 
+  /**
+   * The following validation is performed on the updated schema
+   *
+   * <ul>
+   *   <li>Every field has column IDs and physical names, and that these are unique
+   *   <li>Array/map types have valid nested field IDs defined if icebergCompatV2 is enabled
+   *   <li>Physical column names are preserved across the update
+   * </ul>
+   *
+   * @param currentColumnPathToField column path to field for current schema
+   * @param newColumnPathToField column path to field for updated schema
+   * @param icebergCompatV2 Iceberg compatV2 is enabled
+   * @throws IllegalArgumentException in case of validation failure
+   */
+  static void validateColumnIds(
+      Map<String, StructField> currentColumnPathToField,
+      Map<String, StructField> newColumnPathToField,
+      boolean icebergCompatV2) {
+
+    Map<Long, String> currentFieldIdToPhysicalName = new HashMap<>();
+    Map<Long, List<NestedIdEntry>> currentFieldIdToNestedIds = new HashMap<>();
+
+    for (Map.Entry<String, StructField> entry : currentColumnPathToField.entrySet()) {
+      StructField field = entry.getValue();
+      long columnId = getColumnId(field);
+      currentFieldIdToPhysicalName.put(columnId, getPhysicalName(field));
+
+      // If there are nested IDs save them for subsequent consistency validation
+      if (hasNestedColumnIds(field)) {
+        currentFieldIdToNestedIds.put(columnId, nestedIdEntriesForField(field));
+      }
+    }
+
+    Map<Long, String> updatedFieldIdToPhysicalName = new HashMap<>();
+    Map<Long, List<NestedIdEntry>> updatedFieldIdToNestedIds = new HashMap<>();
+    for (Map.Entry<String, StructField> entry : newColumnPathToField.entrySet()) {
+      StructField field = entry.getValue();
+      validateFieldHasColumnIdAndPhysicalName(field);
+      long columnId = getColumnId(field);
+      // Check that it's not a duplicate column ID
+      if (updatedFieldIdToPhysicalName.containsKey(columnId)) {
+        throw new IllegalArgumentException(
+            String.format("Field %s with id %d already exists", field.getName(), columnId));
+      }
+
+      String physicalName = getPhysicalName(field);
+      String existingPhysicalName = currentFieldIdToPhysicalName.get(columnId);
+
+      // Check that the physical name for a given column ID hasn't changed
+      if (existingPhysicalName != null && !existingPhysicalName.equals(physicalName)) {
```

Comment:
in what case might `existingPhysicalName` be null? If we enable column mapping in this txn AND update the schema? I think we should disallow this right? Or how would we know what the schema changes are if the previous schema doesn't have physical names and such... Not sure, any thoughts?

Reply:
Makes sense! Can you add a comment that that's why it would be null please? not intuitive to a reader

Reply:
But to be clear, when column mapping mode is `name` we use the physical name to determine whether columsn are new or existing right?....

Reply:
Also still curious on whether we support enabling column mapping AND schema evolution in the same txn? This seems unsafe to me?